sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/JsonToRow.java

```diff
@@ -0,0 +1,122 @@
+ *
+ * <p>Currently supported {@link Schema} field types are: <ul> <li>{@link Schema.TypeName#BYTE}</li>
+ * <li>{@link Schema.TypeName#INT16}</li> <li>{@link Schema.TypeName#INT32}</li> <li>{@link
+ * Schema.TypeName#INT64}</li> <li>{@link Schema.TypeName#FLOAT}</li> <li>{@link
+ * Schema.TypeName#DOUBLE}</li> <li>{@link Schema.TypeName#BOOLEAN}</li> <li>{@link
+ * Schema.TypeName#STRING}</li> </ul>
+ *
+ * <p>For specifics of JSON deserialization see {@link RowJsonDeserializer}.
+ *
+ * <p>Conversion is strict, with minimal type coercion:
+ *
+ * <p>Booleans are only parsed from {@code true} or {@code false} literals, not from {@code "true"}
+ * or {@code "false"} strings or any other values (exception is thrown in these cases).
+ *
+ * <p>If a JSON number doesn't fit into the corresponding schema field type, an exception is be
+ * thrown. Strings are not auto-converted to numbers. Floating point numbers are not auto-converted
+ * to integral numbers. Precision loss also causes exceptions.
+ *
+ * <p>Only JSON string values can be parsed into {@link TypeName#STRING}. Numbers, booleans are not
+ * automatically converted, exceptions are thrown in these cases.
+ *
+ * <p>If a schema field is missing from the JSON value, an exception will be thrown.
+ *
+ * <p>Explicit {@code null} literals are allowed in JSON objects. No other values are parsed into
+ * {@code null}.
+ */
+public class JsonToRow {
+
+  public static PTransform<PCollection<? extends String>, PCollection<Row>> withSchema(
+      Schema rowSchema) {
+    return JsonToRowFn.forSchema(rowSchema);
+  }
+
+  static class JsonToRowFn extends PTransform<PCollection<? extends String>, PCollection<Row>> {
+    private transient volatile @Nullable ObjectMapper objectMapper;
+    private Schema schema;
+
+    static JsonToRowFn forSchema(Schema rowSchema) {
+      return new JsonToRowFn(rowSchema);
+    }
+
+    private JsonToRowFn(Schema schema) {
+      this.schema = schema;
+    }
+
+    @Override
+    public PCollection<Row> expand(PCollection<? extends String> jsonStrings) {
+      return jsonStrings
+          .apply(
+              ParDo.of(
+                  new DoFn<String, Row>() {
+                    @ProcessElement
+                    public void processElement(ProcessContext context) {
+                      context.output(jsonToRow(context.element()));
+                    }
+                  }))
+          .setCoder(schema.getRowCoder());
+    }
+
+    private Row jsonToRow(String jsonString) {
+      try {
+        return objectMapper().readValue(jsonString, Row.class);
+      } catch (IOException e) {
+        throw new IllegalArgumentException("Unable to parse json object: " + jsonString, e);
+      }
+    }
+
+    private ObjectMapper objectMapper() {
+      if (this.objectMapper == null) {
+        synchronized (this) {
+          if (this.objectMapper == null) {
+            this.objectMapper = newObjectMapperWith(RowJsonDeserializer.forSchema(this.schema));
```

Comment:
Just curious - racy lazy initialization looks fine here, and if it double initializes you would just get GC'd. Is the library not safe? I kind of recall that it might not be, but I have forgotten the details.

Reply:
It is safe, this initialization logic is mainly to avoid serializing . Right now I'm creating an `ObjectMapper` per `PTransform` instance, each of which corresponds to a single `Schema` which we have to keep as well. At think it is excessive to have either `ObjectMapper` serialized or have more complex logic for initialization, accepting that this happens rarely (hopefully once at construction time for a lot of cases). Do you have  thoughts about it?

Reply:
I think I meant something much less smart that what you are saying. My question is: does it work without any `synchronized` at all?

Reply:
But it doesn't matter so much.